### DISCUSSION TITLE
Add (Steam) achievements

### DISCUSSION THREAD

**Discussion by serano01** [SuperTux/supertux#2064] [Ideas]
https://steamcommunity.com/app/1572920/discussions/0/3193614254329940613/

Achievements are very popular and would make sense when we're heading to Steam. SuperTuxKart for example has an in-game achievement system, which can be checked per main menu. Some ideas for achievements:

- Bad Bad Guy (Kill 100 BadGuys)
- One of every kind (kill ten different BadGuys)
- No more, Mr. Ice Guy (Defeat the Yeti)
- Spring Fever (reach the forest)

etc.

**Comment by Semphriss** [comment]
Considerations:
- Steam offers only a proprietary library. It is risky to use it on a GPL app, although as Steam documentation says, we can "decide" that the game merely communicates with the library; but we should be ready to support that. I already have designs in mind that would keep confusion away best, but they imply developing 1) a library-agnostic achievements system, 2) a completely separate process, perhaps permissively licensed or ["GPL-with-an-exception"](https://www.gnu.org/licenses/gpl-faq.html#GPLIncompatibleLibs) and 3) make the achievements system, SuperTux-side, exposed and available to read by any program, with an openly-documented structure [simple enough not to be considered "complex data structures"](https://www.gnu.org/licenses/gpl-faq.html#GPLPlugins).
- I made Steam builds follow releases; this might either have to wait until the next release, or I will have to make an exception and branch out a "Steam" version that will receive only bugfixes.

**Reply by basxto** [reply]
There is already [steamshim](https://github.com/icculus/steamshim) (zlib licensed), which tries to handle the Steamworks SDK communication in a way compatible with GPL.

It would probably be a good idea to have an achievement interface ingame and support them per world, so that addons can add their own. But only achievements from the storyline should be communicated with steam.

**Reply by Semphriss** [reply]
Has this been challenged legally? The amount of data that transfers between processes seem complex enough to be considered "one work" by the GPL...

**Reply by basxto** [reply]
I doubt that, like with with most GPL things. This FAQ itself is just FSF’s preferred interpretation. But best ask icculus.

https://www.gnu.org/licenses/gpl-faq.html#MereAggregation is probably the more important section.

**Comment by Alasdairbugs** [comment]
i think if we're adding achievements to steam version we should add it to the main version as well for consistency sake
